src/nbl/ext/DebugDraw/CDraw3DLine.cpp

```diff
@@ -80,59 +62,33 @@ CDraw3DLine::CDraw3DLine(const core::smart_refctd_ptr<video::ILogicalDevice>& de
 		assert(rpindependent_pipeline);
 	}
 
-	{
-		video::IGPUGraphicsPipeline::SCreationParams gp_params;
-		gp_params.rasterizationSamplesHint = asset::IImage::ESCF_1_BIT;
-		gp_params.renderpass = core::smart_refctd_ptr<video::IGPURenderpass>(renderpass);
-		gp_params.renderpassIndependent = rpindependent_pipeline;
-		gp_params.subpassIx = 0u;
-
-		m_pipeline = device->createGPUGraphicsPipeline(nullptr, std::move(gp_params));
-	}
 	asset::SBufferBinding<video::IGPUBuffer> bindings[video::IGPUMeshBuffer::MAX_ATTR_BUF_BINDING_COUNT];
 	bindings[0u] = { 0u,core::smart_refctd_ptr<video::IGPUBuffer>(m_device->getDefaultUpStreamingBuffer()->getBuffer()) };
 	m_meshBuffer = core::make_smart_refctd_ptr<video::IGPUMeshBuffer>(std::move(rpindependent_pipeline), nullptr, bindings, asset::SBufferBinding<video::IGPUBuffer>{});
 	m_meshBuffer->setIndexType(asset::EIT_UNKNOWN);
 	m_meshBuffer->setIndexCount(2);
 }
 
-void CDraw3DLine::recordToCommandBuffer()
+void CDraw3DLine::recordToCommandBuffer(video::IGPUCommandBuffer* cmdBuffer, video::IGPUGraphicsPipeline* graphics_pipeline)
 {
-	for (uint32_t i = 0u; i < m_scImageCount; ++i)
-	{
-		auto cb = m_commandBuffers[i];
-		auto fb = m_framebuffers[i];
-		cb->reset(0);
-		cb->begin(0);
-
-		auto* buf = m_meshBuffer->getVertexBufferBindings()->buffer.get();
-		size_t offset = 0u;
-		cb->bindVertexBuffers(0u, 1u, &buf, &offset);
-		cb->pushConstants(const_cast<nbl::video::IGPUPipelineLayout*>(m_meshBuffer->getPipeline()->getLayout()), asset::ISpecializedShader::ESS_VERTEX, 0, sizeof(m_viewProj), &m_viewProj);
-		cb->bindGraphicsPipeline(m_pipeline.get());
-		video::IGPUCommandBuffer::SRenderpassBeginInfo info;
-		asset::VkRect2D area;
-		area.offset = { 0, 0 };
-		area.extent = VkExtent2D{ uint32_t(m_imgSize.X), uint32_t(m_imgSize.Y) };
-		info.renderpass = core::smart_refctd_ptr<video::IGPURenderpass>(m_renderpass);
-		info.framebuffer = core::smart_refctd_ptr<video::IGPUFramebuffer>(fb);
-		info.clearValueCount = 0u;
-		info.clearValues = nullptr;
-		info.renderArea = area;
-		cb->beginRenderPass(&info, asset::ESC_INLINE);
-		cb->draw(m_lines.size() * 2, 1u, 0u, 0u);
-		cb->endRenderPass();
-
-		cb->end();
-	}
+
+	auto cb = cmdBuffer;
+	assert(cb->getState() == IGPUCommandBuffer::ES_RECORDING);
+	auto* buf = m_meshBuffer->getVertexBufferBindings()->buffer.get();
+	size_t offset = 0u;
+	cb->bindVertexBuffers(0u, 1u, &buf, &offset);
+	cb->pushConstants(const_cast<nbl::video::IGPUPipelineLayout*>(m_meshBuffer->getPipeline()->getLayout()), asset::ISpecializedShader::ESS_VERTEX, 0, sizeof(m_viewProj), &m_viewProj);
+	cb->bindGraphicsPipeline(graphics_pipeline);
+	cb->draw(m_lines.size() * 2, 1u, 0u, 0u);
+
 }
 
-void CDraw3DLine::draw(video::IGPUSemaphore* waitSem, video::IGPUSemaphore* signalSem, uint32_t imgNum)
+void CDraw3DLine::updateMeshBuffer()
```

Comment:
don't keep vertex data in upstream buff, have dedicated device local buff for that (recreate if new lines data doesn't fit in old buffer). 
Also make it return via pointers (write to if not nullptr) a fence and a semaphore to wait for copy to happen

Also just a note: multi_free() is horribly wrong here